### ISSUE TITLE
non-normalized or out of bounds coordinate labels for all the labels

### ISSUE THREAD

**Issue by AsimFayyazRaja** [ultralytics/yolov5#7135]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I am training yolov5 on my custom dataset and am getting the non-normalized labels' error. 
I understand that this question has been asked here but I tried its solution and it is not working.

The annotations have x,y, and w,h which means that the bounding box is present from **(x,y) to (x+w,y+h)**. 
I am normalizing them according to Yolov5's requirements like this:

<img width="597" alt="image" src="https://user-images.githubusercontent.com/28213136/160016950-701eec54-cd23-4b20-a884-cb1fa54896e2.png">

The real and normalized annotations are:

<img width="415" alt="image" src="https://user-images.githubusercontent.com/28213136/160017878-2e5fb635-817f-48e5-bbe0-fca1078a0db4.png">

I am using the cv2 rectangle function to display the bounding boxes on the image and it is creating the perfect bounding boxes as displayed in the picture below:
 
`cv2.rectangle(temp_img,(int(x), int(y)),(int(x+w), int(y+h)),color=(0, 255, 0),thickness=2)`
![image](https://user-images.githubusercontent.com/28213136/160017048-4610cdfa-b23b-4900-83bf-c317baff5518.png)

The URL of the image is given below, you can download it and even draw the annotations on it to verify.
[image](https://www.nottingham.ac.uk/Estates/Images-Multimedia/WebImages/University-Card-Student.jpg
)

I am using the below-given command to train the YOLOv5 model on Google Colab.

`!python3 train.py --img 640 --batch 8 --epochs 40 \
  --data ./data/roi.yaml --cfg ./models/yolov5x.yaml \
  --name yolov5x_roi --cache`

Can you please tell me what am I doing wrong in normalizing the coordinates? I have verified that the width, height, and annotations of the image are correct i.e. they lie within the image's boundary.

**Comment by glenn-jocher** [comment]
@AsimFayyazRaja 👋 Hello! Thanks for asking about YOLOv5 🚀 **dataset formatting**. To train correctly your data must be in YOLOv5 format. Please see our [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) tutorial for full documentation on dataset setup and all steps required to start training your first model. A few excerpts from the tutorial:

### 1.1 Create dataset.yaml

[COCO128](https://www.kaggle.com/datasets/ultralytics/coco128) is an example small tutorial dataset composed of the first 128 images in [COCO](http://cocodataset.org/#home) train2017. These same 128 images are used for both training and validation to verify our training pipeline is capable of overfitting. [data/coco128.yaml](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), shown below, is the dataset config file that defines 1) the dataset root directory `path` and relative paths to `train` / `val` / `test` image directories (or *.txt files with image paths), 2) the number of classes `nc` and 3) a list of class `names`:
```yaml
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: ../datasets/coco128  # dataset root dir
train: images/train2017  # train images (relative to 'path') 128 images
val: images/train2017  # val images (relative to 'path') 128 images
test:  # test images (optional)

# Classes
nc: 80  # number of classes
names: [ 'person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
         'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
         'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
         'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
         'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
         'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
         'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard', 'cell phone',
         'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear',
         'hair drier', 'toothbrush' ]  # class names
```


### 1.2 Create Labels

After using a tool like [Roboflow Annotate](https://roboflow.com/annotate?ref=ultralytics) to label your images, export your labels to **YOLO format**, with one `*.txt` file per image (if no objects in image, no `*.txt` file is required). The `*.txt` file specifications are:

- One row per object
- Each row is `class x_center y_center width height` format.
- Box coordinates must be in **normalized xywh** format (from 0 - 1). If your boxes are in pixels, divide `x_center` and `width` by image width, and `y_center` and `height` by image height.
- Class numbers are zero-indexed (start from 0).

<img width="800" alt="Image Labels" src="https://user-images.githubusercontent.com/26833433/91506361-c7965000-e886-11ea-8291-c72b98c25eec.jpg">

The label file corresponding to the above image contains 2 persons (class `0`) and a tie (class `27`):

<p align="center"><img width="428" src="https://user-images.githubusercontent.com/26833433/112467037-d2568c00-8d66-11eb-8796-55402ac0d62f.png"></p>


### 1.3 Organize Directories

Organize your train and val images and labels according to the example below. YOLOv5 assumes  `/coco128` is inside a `/datasets` directory **next to** the `/yolov5` directory. **YOLOv5 locates labels automatically for each image** by replacing the last instance of `/images/` in each image path with `/labels/`. For example:
```bash
../datasets/coco128/images/im0.jpg  # image
../datasets/coco128/labels/im0.txt  # label
```

<p align="center"><img width="700" src="https://user-images.githubusercontent.com/26833433/134436012-65111ad1-9541-4853-81a6-f19a3468b75f.png"></p>
</details>


Good luck 🍀 and let us know if you have any other questions!

**Comment by AsimFayyazRaja** [comment]
@glenn-jocher thanks for the detailed information. I have followed the exact same procedure. My images are located in the correct format. The directory structure and YAML file are shown below.

<img width="375" alt="image" src="https://user-images.githubusercontent.com/28213136/160022409-1849d952-9fbd-48f5-a86f-4e6c30f3ded9.png">

<img width="726" alt="image" src="https://user-images.githubusercontent.com/28213136/160022702-c735f2e3-40c3-4278-b54e-033e0f5354bc.png">


Can you please tell me what am I doing wrong in normalizing the x,y, and w,h coordinates because that is what the assertion is referring to? Thanks!

**Comment by glenn-jocher** [comment]
@AsimFayyazRaja You're doing great! The annotations and file hierarchy look good. The issue might be with the normalization. The YOLOv5 requires box coordinates to be in normalized xywh format (x_center, y_center, width, height), ranging from 0 - 1.

You can normalize your coordinates as follows:
- `x_center_normalized = x / image_width`
- `y_center_normalized = y / image_height`
- `width_normalized = w / image_width`
- `height_normalized = h / image_height`

Be sure to recheck the normalization process and verify that the values are within the required range. For any questions related to data preparation, feel free to consult the YOLOv5 [documentation](https://docs.ultralytics.com/yolov5/) or reach out for further assistance. Good luck with your training!

**Comment by leo-smi** [comment]
For those who wants to convert non-normalized annotations:

```
# xmax ymax xmin ymin -> x_center y_center width height
WIDTH    = 640    # annotated image width
HEIGHT   = 640    # annotated image height
width    = row['xmax'] - row['xmin']
height   = row['ymax'] - row['ymin']
x_center = row['xmin'] + (width  / 2)
y_center = row['ymin'] + (height / 2)

content = f"{class} {x_center / WIDTH} {y_center / HEIGHT} {width / WIDTH} {height / HEIGHT }"  # normalized
```

**Comment by leo-smi** [comment]
I made a visual diagram for helping people understand the normalization for YOLO

![Normalize-Annotations](https://github.com/user-attachments/assets/4cf21314-d99c-45c6-a3d0-244511620ec4)

**Comment by glenn-jocher** [comment]
@leo-smi thank you for sharing the visual diagram! It’s a great resource for understanding the normalization process for YOLO annotations. Visual aids like this can be incredibly helpful for the community.

For those who are looking to convert non-normalized annotations to the required YOLO format, here's a quick code snippet that might help:

```python
# Example conversion from non-normalized (xmax, ymax, xmin, ymin) to normalized (x_center, y_center, width, height)

WIDTH = 640  # annotated image width
HEIGHT = 640  # annotated image height

# Assuming row contains your bounding box coordinates
width = row['xmax'] - row['xmin']
height = row['ymax'] - row['ymin']
x_center = row['xmin'] + (width / 2)
y_center = row['ymin'] + (height / 2)

# Normalize the coordinates
content = f"{class_id} {x_center / WIDTH} {y_center / HEIGHT} {width / WIDTH} {height / HEIGHT}"
```

If you encounter any issues or have further questions, please ensure you are using the latest versions of `torch` and the YOLOv5 repository. If the problem persists, providing a minimum reproducible code example would be very helpful for us to investigate further. You can find more details on how to create one [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Thanks again for contributing to the community! 🚀
